src/GraphQL/Caching/IDocumentCache.cs

```diff
@@ -8,8 +9,18 @@ namespace GraphQL.Caching
     public interface IDocumentCache
     {
         /// <summary>
-        /// Gets or sets a document in the cache. Must be thread-safe. Returns <see langword="null"/> if no entry is found.
+        /// Gets a document in the cache. Must be thread-safe. Returns <see langword="null"/> if no entry is found.
         /// </summary>
-        Document? this[string query] { get; set; }
+        ValueTask<Document?> GetAsync(string query);
+
+        /// <summary>
+        /// Sets a document in the cache. Must be thread-safe.
+        /// </summary>
+        ValueTask SetAsync(string query, Document value);
+
+        /// <summary>
+        /// Remove a document cache. Must be thread-safe.
+        /// </summary>
+        ValueTask RemoveAsync(string query);
```

Comment:
Why do we need this method here? This is an interface (contract) between GraphQL.NET and some external code. GraphQL.NET needs only to get document from cache or set document into cache. That's all.

Reply:
The value arg of the SetAsync method is non-nullable now. There is no longer a valid way to remove a cache item (or set its value to null, at least). I'm not sure if the RemoveAsync is needed, but I reserved it.

I can revert it if you're sure.

Reply:
I think what @sungam3r is saying is that the interface is for GraphQL.NET to use, and GraphQL.NET does not currently need to clear data from the cache.  However, it would make sense for implementations, including MemoryCache, to have a Remove / RemoveAsync method.

@sungam3r Are you sure we should not include a method in the interface for clearing entries?  What about for a scenario where someone wanted to write custom middleware that removes cache entries when, for example, an authentication token has expired?

I can't think of a specific reason myself, but I might include the method anyway, perhaps with a note that its implementation is not required for GraphQL.NET.

Reply:
If one needs `Remove` then he inherits from `MemoryDocumentCache` or introduce another interface containing `Remove`.  It makes no sense to add this method in `IDocumentCache` interface.

Reply:
Even `SetAsync` **may** be implemented in a way that it actually removes supplied key from cache if `value` is `null`.

Reply:
Should I roll it back from
```csharp
Task SetAsync(string query, Document value);
```
to
```csharp
Task SetAsync(string query, Document? value);
```
?

Reply:
I think not:
```
ValueTask SetAsync(string query, Document value);
```

Reply:
> Even `SetAsync` **may** be implemented in a way that it actually removes supplied key from cache if `value` is `null`.

But the value is considered never to be null since the project is `<Nullable>enable</Nullable>`, at least the IDEs do.

Reply:
Yes. I said that `SetAsync` **may** be implemented such way but not really **should**. GraphQL.NET needs `SetAsync` to set not null document. Nullable annotations help to understand that. However, it is technically nothing preventing pass `null!`.

Reply:
If the argument is passed as non-nullable, it should be expected that it may throw an ArgumentNullException if it is null. We should not disguise the method intent behind incorrect NRT annotations.

Reply:
I'm fine either way.